service/game.ts

```diff
@@ -50,11 +49,11 @@ export async function welcome(): Promise<string[]> {
   }
 }
 
-export async function allGamesParsed(): Promise<Result[]> {
+export async function allGamesParsed(): Promise<Error | ResultArray> {
```

Comment:
tendenzialmente in TS è più idiomatico fare `throw` nel caso di errori, generando quindi una `Promise` fallita: perciò il tipo qui può restare solo `Promise<ResultArray>`

e poi fai un `try`/`catch` dove fai `await` di questa promise

nulla vieta di restituire invece un'unione, e noi in progetti passati usavamo per esempio fp-ts con `TaskEither` per usare un `Either` nel risultato, ma ora di solito facciamo `throw` in casi del genere (non rende esplicito l'errore nel tipo, naturalmente, quindi aumenta il rischio di scordarsi di gestirlo, ma in pratica gli errori da gestire esplicitametne sono pochi nelle nostre applicazioni tipiche)

Reply:
Ho scelto di restituire `Promise<Error | ResultArray> ` perchè in questo modo mi viene più semplice estrarre il messaggio di errore dal chiamante. Ho provato anche a lavorare sulla `Promise` fallita, ma il messaggio di errore che mi arriva è quello di zod e non quello a monte. Questo probably è assimilabile all'uso di un `Either` senza importare `fp-ts`, immagino. Mi potresti mostrare uno snippet di come lo scriveresti pls?

Reply:
Tendenzialmente farei così:

1. qui userei `.parse` invece di `.safeParse` e quindi cambierei il tipo di ritorno in `Promise<ResultArray>`
2. dove lo usi farei
    ````
    try {
      const allResults = await allGamesParsed();
      return res
        .status(200)
        .json({
          game_history: allResults,
        })
        .end();
    } catch (error) {
        ...
    }
    ````

Il problema è che `error` avrà tipo `unknown` e non `Error`, perché in JS si può fare throw anche di cose che non sono `Error`. Quindi poi per accedere al campo `.message` devi fare un `instanceof`.

In pratica però l'errore sarà lo stesso che hai nel campo `.error` del risultato di `safeParse`, perché `parse` si limita a usare `safeParse` e poi `throw`. Quindi non dovrebbe essere un messaggio diverso, ma non so se ho capito "il messaggio di errore che mi arriva è quello di zod e non quello a monte".

(In generale mi torna che sia meglio usare `safeParse` per assicurarsi di gestire l'errore, ma in caso come questo, in cui l'errore avviene solo in caso lo schema del DB è disallineato rispetto al codice, quindi in caso di bug, ci sta fare un `parse` diretto per semplicità.)